firmware/console/binary/tunerstudio_io_serial_ports.cpp

```diff
@@ -11,16 +11,23 @@
 #include "tunerstudio_io.h"
 #include "connector_uart_dma.h"
 
+// These may not be defined due to the HAL, but they're necessary for the compiler to do it's magic
+class UARTDriver;
+class UartDmaTsChannel;
+class UartTsChannel;
+class SerialTsChannel;
+
 #if HAS_PRIMARY
-	#ifdef TS_PRIMARY_UART
-		#if EFI_USE_UART_DMA
-			UartDmaTsChannel primaryChannel(TS_PRIMARY_UART);
-		#else
-			UartTsChannel primaryChannel(TS_PRIMARY_UART);
-		#endif
-	#elif defined(TS_PRIMARY_SERIAL)
-		SerialTsChannel primaryChannel(TS_PRIMARY_SERIAL);
-	#endif
+#ifdef TS_PRIMARY_PORT
+	std::conditional_t<
+		std::is_same_v<decltype(TS_PRIMARY_PORT), UARTDriver>,
```

Comment:
please add a one line comment explaining this C++ magic to less experienced C developers.